src/transaction/flashback.c

```diff
@@ -64,9 +64,97 @@ static volatile int flashback_Threshold_to_remove_archive = 0;	/* If the differe
 								 * the archive log volume can be deleted.
 								 */
 // *INDENT-OFF*
-static std::atomic_bool flashback_Is_active = false;	// the status value that the flashback is processing the request
+static std::atomic_bool flashback_Is_active = false;	// the status value that the flashback is processing the a request
+static std::atomic_bool flashback_In_progress = false;  // the status value that the flashback is in progress
 // *INDENT-ON*
 
+static CSS_CONN_ENTRY *flashback_Current_conn = NULL;	// the connection entry for a flashback requests
+
+/*
+ * flashback_is_duplicated_request - check if the caller is duplicated request for flashback
+ *
+ * return   : duplicated or not
+ */
+
+static bool
+flashback_is_duplicated_request (THREAD_ENTRY * thread_p)
+{
+  CSS_CONN_ENTRY *previous_conn = NULL;
+
+  /* When flashback is first called or exited properly,
+   * flashback_Current_conn is set to NULL */
+  if (flashback_Current_conn == NULL)
+    {
+      return false;
+    }
+
+  /* flashback_Current_conn indicates conn_entry in thread_p, and conn_entry can be reused by request handler.
+   * So, status in flashback_Current_conn can be overwritten. */
+
+  previous_conn = flashback_Current_conn;
+
+  if (flashback_In_progress == false)
+    {
+      /* previous flashback set flashback_In_progress to false prolperly. (exited well) */
+      return false;
+    }
+  else
+    {
+      /* flashback_In_progress == true */
+
+      if (previous_conn->in_flashback == true)
+	{
+	  /* previous flashback is still in progress */
+	  return true;
+	}
+      else
+	{
+	  /* - previous_conn is overwritten with new connection, so in_flashback value is initialized to false.
+	   * - previous flashback connection has been exited abnormally,
+	   *   so existing flashback_In_progress value has not been initialized */
+	  return false;
+	}
+    }
+}
+
+/*
+ * flashback_initialize - check request if is duplicated, 
+ *                        then initialize variables and connection when flashback request is started
+ */
+
+int
+flashback_initialize (THREAD_ENTRY * thread_p)
+{
+  static pthread_mutex_t conn_mutex = PTHREAD_MUTEX_INITIALIZER;
```

Comment:
I'm still not sure whether global is better or local is. The target of this mutex is global, so it seems to have to be on the same level (as a global var), but it is used only here. Now I think it as a global var is better, but I'll follow your choice.

Reply:
Yes, I agree with what you said. 
I think it's better to set this as a global variable according to the target of this lock.